### PR TITLE
KAFKA-8347: Choose next record to process by timestamp

### PR DESCRIPTION
**Pull request by ableegoldman**
When choosing the next record to process, we should look at the head record's timestamp of each partition and choose the lowest rather than choosing the lowest of the partition's streamtime.

This change effectively makes RecordQueue return the timestamp of the head record rather than its streamtime. Streamtime is removed (replaced) from RecordQueue as it was only being tracked in order to choose the next partition to poll from.

### CODE REVIEW

**Review by mjsax — APPROVED** [review]
LGTM -- not sure why Jenkins failed.

Retest this please.

**Comment by mjsax** [comment]
Retest this please.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java]: I noticed the comments throughout this test were not accurate, so I fixed them to reflect the current streamtime logic

**Comment by ableegoldman** [comment]
I refactored this somewhat to return RecordQueue.UNKNOWN from timestamp() when the queue is empty, as timestamp() should now be reflecting not stream time but timestamp of the next (head) record. Also fixed some out of date/inaccurate comments in PartitionGroupTest

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java]: nit: should we merge this into existing test? If not, rename to `shouldChooseNextRecordBasedOnHeadTimestampe`

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java]: nit: there is not "partition stream time" -- there is "stream time" and "partition time" :)

Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java]: Ack, I chose to use the term to differentiate between the "old" partition time -- largest timestamp observed for a partition so far -- and what is effectively becoming the "new" partition time in this PR, ie timestamp of head record. Will clarify the comment though

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java]: Ack

**Comment by ableegoldman** [comment]
Closing this in favor of [minor version bump](https://github.com/apache/kafka/pull/6743)

**Review by mjsax — APPROVED** [review]
LGTM.

Call for second review @guozhangwang @bbejeck @vvcephei @cadonna @abbccdda

**Review by bbejeck — APPROVED** [review]
LGTM

**Comment by bbejeck** [comment]
Merged #6719 into trunk.

**Comment by bbejeck** [comment]
cherry-picked to 2.2 and 2.1

**Pull request merged**
